### PR TITLE
Adding ability to index by multiple properties

### PR DESCRIPTION
**Pull request by atrites**
Using the schema.index() function we can now pass an array to create a compound index. This will result in properties being index by all fields passed.

eg - 

schema.index(['email', 'username']); 

will index the document as $ref_by_email_and_username_{$email}_{$username}

### CODE REVIEW

File: lib/cbdocument.utils.js [hunk 1]
```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
...
-            if (typeof v === 'string') {
-              return v;
-            }
-            if (typeof v === 'number') {
-              return v.toString();
-            }
-            if (typeof v === 'object' && v._isBaseModel) {
-              return v.getDocumentKeyValue(v.config.storeFullReferenceId);
+
+      const processValue = initialValue => {
```
Comment (bojand) [inline lib/cbdocument.utils.js]: if this doesn't need to be here due to lexical closure or something else (from initial look it doesn't seem to be) I would say just make it into a separate function within utils.

File: lib/cbdocument.utils.js [hunk 2]
```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
...
+          finalVal = null;
+        }
+        return finalVal;
+      }
+
+      let value = null;
+      if (indexes[v].compound) {
+        value = _.map(path, p => {
+          return processValue(mpath.get(p, data));
         });
```
Comment (bojand) [inline lib/cbdocument.utils.js]:
``` js
value = _.map(path, p => processValue(mpath.get(p, data))); 
```

should work and looks nicer to me

File: lib/cbdocument.utils.js [hunk 3]
```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
...
 
-        valToStore = _.compact(valToStore);
-        valToStore = valToStore.sort();
-      } else if (_.isUndefined(value) || _.isNull(value)) {
-        valToStore = null;
+        const expandValues = valueList => {
```
Comment (bojand) [inline lib/cbdocument.utils.js]: same here if there is no reason for it to be within this function, make it separate.

File: lib/cbdocument.utils.js [hunk 4]
```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
...
+          var retVal = [];
+          _.each(valueList, val => {
+            if (Array.isArray(val)) {
+              let expVal = expandValues(val);
+              let intRet = [];
```
Comment (bojand) [inline lib/cbdocument.utils.js]: this should be `const`? was linter (xo) passing w/o errors? also meaning behind the variable name `intRet`?

File: lib/cbdocument.utils.js [hunk 5]
```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
...
+              retVal.push(val);
+            }
+          });
+          return retVal;
+        }
+
+        value = expandValues(value);
+        value = _.compact(value);
+
+        let containsNested = _.some(value, _.isArray);
```
Comment (bojand) [inline lib/cbdocument.utils.js]: `const`

File: lib/cbdocument.utils.js [hunk 6]
```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
...
+        if (containsNested) {
+          value = _.map(value, v => {
+            return v.join('_');
+          });
```
Comment (bojand) [inline lib/cbdocument.utils.js]:
``` js
value = _.map(value, v => v.join('_'));
```

File: lib/cbdocument.utils.js [hunk 7]
```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
...
+        }
+
+      } else {
+        value = processValue(mpath.get(path, data));
       }
 
       ret[name] = {
         path,
-        value: valToStore,
+        value: value,
```
Comment (bojand) [inline lib/cbdocument.utils.js]: just `value` should do

``` js
ret[name] = {
  path,
  value,
  name,
  indexType
};
```

File: lib/cbdocument.utils.js [hunk 8]
```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
...
+        value = _.map(path, p => {
+          return processValue(mpath.get(p, data));
         });
 
-        valToStore = _.compact(valToStore);
-        valToStore = valToStore.sort();
-      } else if (_.isUndefined(value) || _.isNull(value)) {
-        valToStore = null;
+        const expandValues = valueList => {
+          var retVal = [];
```
Comment (jcurtis) [inline lib/cbdocument.utils.js]: Embrace the `let` and the `const`

File: lib/schema.js [hunk 9]
```diff
@@ -187,16 +188,39 @@ export default class Schema {
...
+      _.each(prop, p => {
+        if (!this.descriptor[p]){
+          throw new Error(`Descriptor not found for ${p}`);
+        }
+      });
+
+      let indexType = options.indexType ? options.indexType: 'single';
+      let indexName = options.indexName;
+      if (!indexName){
+        const reg = new RegExp('^\\d+$');
```
Comment (jcurtis) [inline lib/schema.js]: Define this outside the function and loop scope if it's going to be reused.

File: lib/schema.js [hunk 10]
```diff
@@ -187,16 +188,39 @@ export default class Schema {
...
+    } else {
+      if (!this.descriptor[prop]){
+        throw new Error('Descriptor not found for this field');
+      }
 
-    let fullDes = {};
-    fullDes[prop] = this.descriptor[prop];
-    this._getIndexes(fullDes);
+      this.descriptor[prop].index = true;
+      this.descriptor[prop] = _.extend(this.descriptor[prop], options);
```
Comment (jcurtis) [inline lib/schema.js:219]: `_.extend` mutates the object so I'd do `_.extend({}, this.descriptor[prop], options);`

**Comment by jcurtis** [comment]
By "multiple" do you mean just two properties or more than two? Should probably add some tests for several if it's the later case.

File: lib/schema.js [hunk 11]
```diff
@@ -187,16 +190,38 @@ export default class Schema {
    *
    */
   index(prop, options = {}){
```
Comment (jcurtis) [inline lib/schema.js:192]: Few spots with `){`, should be `) {`. I blame Bojan.

File: lib/utils.js [hunk 12]
```diff
@@ -319,3 +319,53 @@ export function promisifyCall(ctx, fn, ...args) {
     fn.apply(ctx, args);
   });
 }
+
+export function expandValues(valueList) {
```
Comment (jcurtis) [inline lib/utils.js:323]: Comments on these utils would be nice when you clean everything up for the real PR

Comment (atrites) [reply lib/utils.js:323]: 👍

File: lib/schema.js (hunk 11 above)
Comment (atrites) [reply lib/schema.js:192]: 😠

**Comment by jcurtis** [comment]
🍍

**Pull request merged**